src/ui/popup/regex.tsx

```diff
@@ -0,0 +1,188 @@
+import {
+	Match,
+	Resource,
+	Show,
+	Switch,
+	createEffect,
+	createSignal,
+} from 'solid-js';
+import {
+	EditedFields,
+	FieldType,
+	RegexFields,
+	getProcessedFieldsNoRegex,
+	getSongFieldNoRegex,
+	pascalCaseField,
+	replaceFields,
+	searchMatches,
+} from '@/util/regex';
+import { ManagerTab } from '@/core/storage/wrapper';
+import regexEdits from '@/core/storage/regex-edits';
+import { sendBackgroundMessage } from '@/util/communication';
+
+const [searches, setSearches] = createSignal({
+	track: null,
+	artist: null,
+	album: null,
+	albumArtist: null,
+} as RegexFields);
+const [replaces, setReplaces] = createSignal({
+	track: null,
+	artist: null,
+	album: null,
+	albumArtist: null,
+} as RegexFields);
+const [previews, setPreviews] = createSignal({
+	track: '',
+	artist: '',
+	album: '',
+	albumArtist: '',
+} as EditedFields);
+
+/**
+ * Regex editing popup
+ */
+export default function Regex(props: {
+	clonedSong: ClonedSong;
+	tab: Resource<ManagerTab>;
+}) {
+	return (
+		<div class={styles.regexContainer}>
+			<span class={styles.searchLabel}>{t('infoSearchLabel')}</span>
+			<span class={styles.replaceLabel}>{t('infoReplaceLabel')}</span>
+			<span class={styles.previewLabel}>{t('infoPreviewLabel')}</span>
+			<Entry clonedSong={props.clonedSong} type="track" />
+			<Entry clonedSong={props.clonedSong} type="artist" />
+			<Entry clonedSong={props.clonedSong} type="album" />
+			<Entry clonedSong={props.clonedSong} type="albumArtist" />
+			<Footer tab={props.tab} clonedSong={props.clonedSong} />
+		</div>
+	);
+}
+
+/**
+ * Label, inputs, and preview for a single song field.
+ */
+function Entry(props: { clonedSong: ClonedSong; type: FieldType }) {
+	const { clonedSong, type } = props;
+	setPreviews((prev) => ({
+		...prev,
+		[type]: getSongFieldNoRegex(clonedSong, type),
+	}));
+
+	createEffect(() => {
+		setPreviews(() =>
+			replaceFields(
+				searches(),
+				replaces(),
+				getProcessedFieldsNoRegex(clonedSong)
+			)
+		);
+	});
+
+	return (
+		<>
+			<span class={`${styles[`${type}Label`]} ${styles.entryLabel}`}>
+				{t(`info${pascalCaseField(type)}Label`)}
+			</span>
+			<SearchField type={type} clonedSong={clonedSong} />
+			<input
+				type="text"
+				class={styles[`${type}Replace`]}
+				onInput={(e) =>
+					setReplaces((prev) => ({
+						...prev,
+						[type]: e.currentTarget.value || null,
+					}))
+				}
+			/>
+			<span class={styles[`${type}Preview`]}>{previews()[type]}</span>
+		</>
+	);
+}
+
+/**
+ * A single search input for a single song property.
+ */
+function SearchField(props: { type: FieldType; clonedSong: ClonedSong }) {
+	const { type, clonedSong } = props;
+	return (
+		<div class={`${styles[`${type}Search`]} ${styles.searchWrapper}`}>
+			<input
+				type="text"
+				class={styles.searchField}
+				onInput={(e) =>
+					setSearches((prev) => ({
+						...prev,
+						[type]: e.currentTarget.value || null,
+					}))
+				}
+			/>
+			<Switch
+				fallback={
+					<Close
```

Comment:
Should this unmatched state block saving?

Reply:
I think it should probably just serve as a guide and not block anything personally.

If the user wants to save anyway probably just let them, the extension is pretty clear that the edit should be doing nothing to the current track.